source/crd.go

```diff
@@ -0,0 +1,154 @@
+)
+
+// crdSource is an implementation of Source that provides endpoints by listing
+// specified CRD and fetching Endpoints embedded in Spec.
+type crdSource struct {
+	crdClient   rest.Interface
+	namespace   string
+	crdResource string
+	codec       runtime.ParameterCodec
+}
+
+func addKnownTypes(scheme *runtime.Scheme, groupVersion schema.GroupVersion) error {
+	scheme.AddKnownTypes(groupVersion,
+		&endpoint.DNSEndpoint{},
+		&endpoint.DNSEndpointList{},
+	)
+	metav1.AddToGroupVersion(scheme, groupVersion)
+	return nil
+}
+
+// NewCRDClientForAPIVersionKind return rest client for the given apiVersion and kind of the CRD
+func NewCRDClientForAPIVersionKind(client kubernetes.Interface, kubeConfig, kubeMaster, apiVersion, kind string) (*rest.RESTClient, *runtime.Scheme, error) {
+	if kubeConfig == "" {
+		if _, err := os.Stat(clientcmd.RecommendedHomeFile); err == nil {
+			kubeConfig = clientcmd.RecommendedHomeFile
+		}
+	}
+
+	config, err := clientcmd.BuildConfigFromFlags(kubeMaster, kubeConfig)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	groupVersion, err := schema.ParseGroupVersion(apiVersion)
+	if err != nil {
+		return nil, nil, err
+	}
+	apiResourceList, err := client.Discovery().ServerResourcesForGroupVersion(groupVersion.String())
+	if err != nil {
+		return nil, nil, fmt.Errorf("error listing resources in GroupVersion %q", groupVersion.String())
+	}
+
+	var crdAPIResource *metav1.APIResource
+	for _, apiResource := range apiResourceList.APIResources {
+		if apiResource.Kind == kind {
+			crdAPIResource = &apiResource
+			break
+		}
+	}
+	if crdAPIResource == nil {
+		return nil, nil, fmt.Errorf("unable to find Resource Kind %q in GroupVersion %q", kind, apiVersion)
+	}
+
+	scheme := runtime.NewScheme()
+	addKnownTypes(scheme, groupVersion)
+
+	config.ContentConfig.GroupVersion = &groupVersion
+	config.APIPath = "/apis"
+	config.NegotiatedSerializer = serializer.DirectCodecFactory{CodecFactory: serializer.NewCodecFactory(scheme)}
+
+	crdClient, err := rest.UnversionedRESTClientFor(config)
+	if err != nil {
+		return nil, nil, err
+	}
+	return crdClient, scheme, nil
+}
+
+// NewCRDSource creates a new crdSource with the given config.
+func NewCRDSource(crdClient rest.Interface, namespace, kind string, scheme *runtime.Scheme) (Source, error) {
+	return &crdSource{
+		crdResource: strings.ToLower(kind) + "s",
+		namespace:   namespace,
+		crdClient:   crdClient,
+		codec:       runtime.NewParameterCodec(scheme),
+	}, nil
+}
+
+// Endpoints returns endpoint objects.
+func (cs *crdSource) Endpoints() ([]*endpoint.Endpoint, error) {
+	endpoints := []*endpoint.Endpoint{}
+
+	result, err := cs.List(&metav1.ListOptions{})
+	if err != nil {
+		return nil, err
+	}
+
+	for _, dnsEndpoint := range result.Items {
+		endpoints = append(endpoints, dnsEndpoint.Spec.Endpoints...)
+		dnsEndpoint.Status.ObservedGeneration = dnsEndpoint.Generation
+		// Update the ObservedGeneration
+		_, err = cs.UpdateStatus(&dnsEndpoint)
```

Comment:
We should update the observed generation as we do work - it's not clear to me where this method is used. Am I missing something?

Reply:
`Endpoints()` is the interface method for Source. The controller calls the Source's Endpoint method and proceeds onto making necessary changes to DNS records in provider. Source is a interface with just this `Endpoints()` method.